go/common/crypto/sakg/bip32.go

```diff
@@ -0,0 +1,97 @@
+package sakg
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// HardenedKeysIndexStart is the index of the first hardened BIP-0032 key.
+const HardenedKeysIndexStart = uint32(0x80000000)
+
+// BIP32PathMnemonicComponent is the string representing the mnemonic (first)
+// component of a BIP-0032 path.
+const BIP32PathMnemonicComponent = "m"
+
+// BIP32Path represents a BIP-0032 path.
+type BIP32Path []uint32
+
+// String returns the string representation of a BIP-0032 path.
+//
+// NOTE: Hardened paths are marked with '.
+func (path BIP32Path) String() string {
+	pathStr := "m"
+	for _, component := range path {
+		if component >= HardenedKeysIndexStart {
+			component -= HardenedKeysIndexStart
+			pathStr += fmt.Sprintf("/%d'", component)
+		} else {
+			pathStr += fmt.Sprintf("/%d", component)
+		}
+	}
+	return pathStr
+}
+
+// MarshallText encodes a BIP-0032 path into text form.
+func (path BIP32Path) MarshalText() ([]byte, error) {
+	return []byte(path.String()), nil
+}
+
+// UnmarshalText decodes a text marshaled BIP-0032 path.
+func (path *BIP32Path) UnmarshalText(text []byte) error {
+	components := strings.Split(string(text), "/")
+	// NOTE: The first component should be the mnemonic component which doesn't
+	// have a corresponding element in BIP32Path's slice.
+	n := len(components) - 1
+	rawPath := make([]uint32, n)
+
+	for i, component := range components {
```

Comment:
This whole loop is kind of hard to read.  Probably should do something like:
```
  if len(components) < 1 {
    return fmt.Errorf("nuh uh uh you didn't say the magic word")
  }
  rawPath := make([]uint32, 0, n-1)

  if components[0] != BIP32PathMnemonicComponent {
    return fmt.Errorf("wtf r u doing?")
  }

  for i, component := range components[1:] {
    hardened := strings.HasSuffix(component, BIP32HardenedComponentSuffix)
    if hardened {
      component = strings.TrimSuffix(component, BIP32HardenedComponentSuffix)
    }
    // Blah Blah Blah
    rawPath = append(rawPath, comp32)
  }
```

Reply:
Agreed. Thanks!